### ISSUE TITLE
Content Enhancement : KNN

### ISSUE THREAD

**Issue by AM1CODES** [OneStep-elecTRON/ContentSection#14]
The below mentioned additions are required - 
* Math - We need to add a solved example to show the working of the algorithm. To be precise, a pen and paper example of the math behind the algorithm with help of images like we have used for other algorithms.

**Comment by sudharsansrini** [comment]
### KNN - K Nearest Neighbor

KNN algorithm is mainly used to classify and predicting the values in Machine Learning. We can use it for Supervised Machine Learning problems(Labeled) such that Classification and Regression.

**KNN**  
K - constant value
NN - Nearest Neighbor

**KNN is simply , Finding the K Nearest Neighbor of the data point and classifying the data points with maximum number of classes of the neighbors**

![20210831_155414_1](https://user-images.githubusercontent.com/65889571/131488914-f008b056-85ba-4ea5-9635-bc574fedb335.jpg)

We can find the value of K by finding the mean of the error in each records and plotting as graph
K is the value where the mean error rate is lesser


```
error =[]
for i in range(1, 40):
    model = KNeighborsClassifier(n_neighbors=i)
    model.fit(x_train, y_train)
    pred_i = model.predict(x_test)
    error.append(np.mean(pred_i != y_test))

plt.figure(figsize=(15, 15))
plt.plot(range(1, 40), error, color = "red", linestyle = "dashed", marker="o", markerfacecolor="blue", markersize=10)
plt.title('Error Rate K value')
plt.xlabel('K value')
plt.ylabel('Mean Error')
plt.show()
```

Finding K -> Finding how many neighbors that we have to choose around the data point

![Figure_1](https://user-images.githubusercontent.com/65889571/131488179-fa6b401f-56b8-4291-86fa-ad404bf5984c.png)

Value of k = less error rate
After choosing K value, 
You may have a question like, 
How will it find the k nearest points ? 

Distance of the data points will be calculated in KNN using Minkowski distance.

<img width="138" alt="min" src="https://user-images.githubusercontent.com/65889571/131489996-27b59c0a-ea0f-44ef-adbf-f7621e025c46.png">

Based on Minkowski Distance metric, We gonna classify the data points.
p=1, Manhattan Distance
p=2, Euclidean Distance

### Manhattan Distance

![1_KDgfdK6SooXtaUvlnXdpaA](https://user-images.githubusercontent.com/65889571/131490703-8c6a9f26-b787-4d8b-a0a4-957831dc7c24.png)

This is the distance between real vectors using sum of their absolute difference.

### Euclidean Distance

![1_9LeaMTcOXxeTPN-VCbKloQ](https://user-images.githubusercontent.com/65889571/131491201-913c72c3-d2a4-4240-9acf-36bfc6f14ba9.png)

Euclidean distance is calculated as the square root of sum of the squared differences between a new point (x) & existing point y

### Hamming Distance

![1_1lXpSrDyY37V4Ke--GkRQQ](https://user-images.githubusercontent.com/65889571/131491526-4e422c05-dbc9-4482-9ce0-3224baac2be1.png)

It is used for categorical variables. If the value x and the value y are the same, the distance d will be equal to 0. Otherwise distance equal to 1

The above methods are used to find the distance and to finding the k closest neighbor of that particular data point.
With the classes of neighbors, We can classify the data point with maximum neighbor's class

Thank you!

**Comment by AM1CODES** [comment]
Hey @sudharsansrini the content that you mentioned above looks really amazing and we could incorporate it in the main platform. I am pretty sure that writing this would have taken a good amount of time and i request you to make a PR for the same because otherwise it won't get counted as your contribution. Feel free to join our discord if you need any help with how to contribute and stuff like that.
